fsm/fsm.go

```diff
@@ -0,0 +1,254 @@
+package fsm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// ErrEventRejected is the error returned when the state machine cannot process
+// an event in the state that it is in.
+var (
+	ErrEventRejected        = errors.New("event rejected")
+	ErrWaitForStateTimedOut = errors.New("timed out while waiting for event")
+	ErrInvalidContextType   = errors.New("invalid context")
+)
+
+const (
+	// Default represents the default state of the system.
+	Default StateType = ""
+
+	// NoOp represents a no-op event.
+	NoOp EventType = "NoOp"
+
+	OnError EventType = "OnError"
+
+	ContextValidationFailed EventType = "ContextValidationFailed"
+)
+
+// StateType represents an extensible state type in the state machine.
+type StateType string
+
+// EventType represents an extensible event type in the state machine.
+type EventType string
+
+// EventContext represents the context to be passed to the action
+// implementation.
+type EventContext interface{}
+
+// Action represents the action to be executed in a given state.
+type Action func(eventCtx EventContext) EventType
+
+// Transitions represents a mapping of events and states.
+type Transitions map[EventType]StateType
+
+// State binds a state with an action and a set of events it can handle.
+type State struct {
+	EntryFunc   func()
+	ExitFunc    func()
+	Action      Action
+	Transitions Transitions
+}
+
+// States represents a mapping of states and their implementations.
+type States map[StateType]State
+
+// Notification represents a notification sent to the state machine's
+// notification channel.
+type Notification struct {
+	PreviousState StateType
+	NextState     StateType
+	Event         EventType
+}
+
+// StateMachine represents the state machine.
+type StateMachine struct {
+	// Context represents the state machine context.
+	States States
+
+	// Previous represents the previous state.
+	Previous StateType
+
+	// Current represents the current state.
+	Current StateType
```

Comment:
Given that `SendEvent` (and then `getNextState`) changes the `Current`'s and `Previous`' value we might want to instead expose these as functions that also lock the mutex.